### ISSUE TITLE
When restoring, cannot rewrite files that are readonly

### ISSUE THREAD

**Issue by toh995** [restic/restic#4757]
Output of `restic version`
--------------------------
```
restic 0.16.4 compiled with go1.22.1 on linux/amd64
```

What backend/service did you use to store the repository?
---------------------------------------------------------
Local

Problem description / Steps to reproduce
----------------------------------------
NOTE: the following script should be copy+pasted into bash! (zsh didn't work for me)
```bash
# Initialize a restic repo
export RESTIC_REPOSITORY=/tmp/restic
export RESTIC_PASSWORD=foo
mkdir -p $RESTIC_REPOSITORY
restic init

# Create two new directories
DIR1=/tmp/foo1 && mkdir -p $DIR1
DIR2=/tmp/foo2 && mkdir -p $DIR2

# Create a readonly file in DIR1
# RO == readonly
RO_FILE=readonly.txt
echo "first" > "${DIR1}/${RO_FILE}"
chmod 444 "${DIR1}/${RO_FILE}"

# Create a snapshot of DIR1
cd $DIR1 && restic backup .

# Restore the snapshot to DIR2:
restic restore latest --target $DIR2

# Within DIR2, replace the existing readonly file with a new one with different contents
rm --force "${DIR2}/${RO_FILE}"
echo "second" > "${DIR2}/${RO_FILE}"
chmod 444 "${DIR2}/${RO_FILE}"

# Create a new snapshot of DIR2:
cd $DIR2 && restic backup .

# Here's the error!
# Try to restore the DIR2 snapshot into DIR1
DEBUG_LOG=logfile.txt restic restore latest --target $DIR1
```

Expected behavior
-----------------
1. Successfully restores the new file contents, for the readonly file
2. The restore returns a 0 exit code
3. No error messages displayed

Actual behavior
---------------
1. Does NOT restore the new file contents
2. The restore returns a 1 exit code
3. Error messages are displayed

For (3), here's the output displayed in the terminal, for the final "restore" command, i.e. `DEBUG_LOG=logfile.txt restic restore latest --target $DIR1`
```
debug log file logfile.txt
debug enabled
repository 493cf5ac opened (version 2, compression level auto)

restoring <Snapshot caeea779 of [/tmp/foo2] at ${REDACTED} by ${REDACTED} to /tmp/foo1
ignoring error for /readonly.txt: open /tmp/foo1/readonly.txt: permission denied
Summary: Restored 1 files/dirs (7 B) in 0:00
Fatal: There were 1 errors
```

I've also attached the [debug logs](https://github.com/restic/restic/files/14901276/logfile.txt)

Do you have any idea what may have caused this?
-----------------------------------------------
I'm not certain.

Maybe we should be hitting this codepath, but it's not for some reason?
https://github.com/restic/restic/blob/6091029fd66eb8d35e8d3763996c741e3dbf5c5e/internal/restorer/fileswriter.go#L56-L64

Also - is this related to this [previous issue](https://github.com/restic/restic/issues/656#issuecomment-258602634) re: readonly _directories_? (The current issue is about readonly _files_)

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
It's a cool tool! :smile: 

---
I'm not sure if the maintainers have time to look at this - if not, I am willing to contribute some code to help fix this.

However, I will definitely need some help/guidance. In particular:
- I don't quite understand the code structure. Seems like there are multiple layers of abstraction:
    - `cmd/restic/cmd_restore.go`
    - `Restorer` - seems like this also does some kind of traversal through the file directory tree? 
    - `filesWriter`
- If we were to write a test, is it more appropriate to write an "integration" test i.e. in `cmd/restic/cmd_restore_integration_test.go`? Or, write more of a "unit-level" test?

**Comment by MichaelEischer** [comment]
Fixing this restore issue is planned for restic 0.17.0 as part of properly handling in-place restores, see also https://github.com/restic/restic/issues/4575 .

**Comment by MichaelEischer** [comment]
This has already been fixed by #4611.
